20240129-satellite-managed-encryption.md

```diff
@@ -0,0 +1,171 @@
+
+### Goals
+
+Users are provided the option to create a project that uses "satellite-managed encryption". One of these projects:
+* uses a securely-stored, random passphrase stored on the satellite as the project's encryption passphrase
+* restricts all files to be uploaded without path encryption
+
+The main goals are to:
+* provide users with an experience that does not require managing and entering an encryption passphrase in the Satellite UI
+* provide users with an easy way to view and list files in lexographical order
+
+### Approach / Design
+
+#### Key Management Store
+
+We need a secure way to store and look up project encryption passphrases on the Satellite. Someone with normal Satellite DB access should not be able to view this data, and ideally it is encrypted. At the same time, it must be very reliable, as losing data about project encryption passphrases means that data may have been lost. The primary features and concerns we should keep in mind:
+* as simple as possible from an implementation perspective, e.g. key-value lookup for project ID -> encryption passphrase
+* secure - should be encrypted, and highly restricted who/what has access to it
+* reliable - should have the same standards for uptime/data loss as metainfo DB
+
+The KMS will be made up of 3 components:
+* The master key and store - The master key will be used to encrypt and decrypt passphrases. The store will be used to securely store the master key. The store will be [Google's Secret Manager](https://cloud.google.com/security/products/secret-manager), which is a secure store for secrets such as API keys and other sensitive data.
+This secrets manager was chosen because it is just as secure as the alternatives, if not more and yet cheaper and easier to manage. It also is more convenient to use as we already use GCP.
+* The KMS service - This is part of the Satellite that will be responsible for fetching the master key at startup, encrypting and decrypting passphrases, and storing and retrieving same from the passphrase store. The master key will be kept in memory throughout the lifetime of the Satellite.
+The satellite will be configured with credentials needed to access the Secret Manager (the master key store).
+* The Passphrase store - This where the encrypted passphrases will be stored. It will be a new column, `passphrase_enc` on the `projects` table.
+
+#### Satellite Project Updates
+
+A new flag should be added to the `projects` table, like `projects.satellite_managed_encryption`. This should be false for all existing projects. While satellite-managed-encryption (this design) is in-progress, new projects should be created with this flag set to false.
+Another column should be added to the `projects` table to be the passphrase store, like `projects.passphrase_enc`. This should be the encrypted passphrase for satellite-managed-encryption projects only and null for others.
```

Comment:
Why do we need to store the encrypted passphrase in the satellite DB? It's already stored in Google Secret Manager and can be accessed from there whenever necessary, doesn't it?

Reply:
If only the "master key" is in Google Secret Manager, we just need to retrieve it once, on startup. Then we can decrypt anything in the `projects` table without needing to make any additional network requests - we just use the master key in memory.
Also, by having encrypted passphrase in the projects table, those encrypted passphrases will automatically be included in our Satellite DB backups. We can still back up Google Secret Manager, but with this implementation, we only need to worry about a single key rather than one for every project.

Reply:
As far as I remember, we discussed using the secret project ID as the satellite-managed passphrase. But then we decided to generate random passphrases and store them in a KMS because those with access to the satellite database shouldn't access the passphrases. Now it turns out we want only to store the master encryption key in the KMS and have the encrypted passphrases in the satellite DB...

If we only have the master encryption key in the KMS, how about storing the secret project ID encrypted in the satellite DB instead of in plaintext? And use it as the satellite-managed passphrase? This way, we won't need an additional `passphrase_enc` column.

Reply:
I do not remember talking about using secret project ID in this way (are you just talking about `projects.id`, the primary key? The one that is not `projects.public_id`?)
Anyway, I don't think it matters where exactly the passphrase is stored, as long as:
* the raw passphrase is not accessible with only cockroach DB access
* the raw passphrase can easily be derived and/or decrypted by the satellite, given a secret stored outside of cockroach

It seems like there are several ways to do this. However, if we wanted to involve `project.id` instead of adding a new column like `passphrase_enc`, maybe the way to do it is to combine `project.id` + `KMS master key` (e.g. crypto hash function) to get the encryption passphrase. And it would not need to be stored anywhere, since it could be derived from data that is already stored.

Reply:
I talked with Wilfred about this, and I think there is one potential problem with deriving the passphrase without having an additional column in the DB: rotating the master key.
If, for some reason, we wanted to change the master key in the KMS (e.g. because it was exposed), it is very straightforward if we are storing `passphrase_enc` in the DB:
* unencrypt `passphrase_enc` with old master key
* encrypt `passphrase` to `passphrase_enc` with new master key
* done
However, if the encryption passphrase is derived based on, for example, master key + project ID, the above process of migration is impossible. If the master key is leaked, the encryption passphrase is permanently leaked, because it is fundamentally derived from the master key.

In short:
* if we store `passphrase_enc` in the DB, we can change the master key without changing the underlying encryption passphrase of the project
* if we derive passphrase rather than storing it, we cannot change the master key unless we re-upload all objects in all the projects (which we will not do).

So for security reasons, I think it is preferable to store `passphrase_enc` in the DB

Reply:
Yes, I referred to `projects.id` as the secret project ID. But if it is the primary key, migrating the DB to encrypt it would be tricky.

I see your argument about rotating the master key and it makes sense.